src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Debugging/RazorBreakpointSpanEndpoint.cs

```diff
@@ -0,0 +1,202 @@
+            var languageKind = _documentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            // If we're in C#, then map to the right position in the generated document
+            if (languageKind == RazorLanguageKind.CSharp &&
+                !_documentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out _, out projectedIndex))
+            {
+                return null;
+            }
+            // Otherwise see if there is more C# on the line to map to
+            else if (languageKind == RazorLanguageKind.Html &&
+                !_documentMappingService.TryMapToProjectedDocumentOrNextCSharpPosition(codeDocument, hostDocumentIndex, out _, out projectedIndex))
+            {
+                return null;
+            }
+            else
+            {
+                return null;
+            }
+
+            // Now ask Roslyn to adjust the breakpoint to a valid location in the code
+            var csharpDocument = codeDocument.GetCSharpDocument();
+            var syntaxTree = CSharpSyntaxTree.ParseText(csharpDocument.GeneratedCode, cancellationToken: cancellationToken);
+            if (!RazorBreakpointSpans.TryGetBreakpointSpan(syntaxTree, projectedIndex, cancellationToken, out var csharpBreakpointSpan))
+            {
+                return null;
+            }
+
+            var csharpText = codeDocument.GetCSharpSourceText();
+
+            csharpText.GetLineAndOffset(csharpBreakpointSpan.Start, out var startLineIndex, out var startCharacterIndex);
+            csharpText.GetLineAndOffset(csharpBreakpointSpan.End, out var endLineIndex, out var endCharacterIndex);
+
+            var projectedRange = new Range()
+            {
+                Start = new Position(startLineIndex, startCharacterIndex),
+                End = new Position(endLineIndex, endCharacterIndex),
+            };
+
+            // Now map that new C# location back to the host document
+            var mappingBehavior = GetMappingBehavior(codeDocument.Source.RelativePath);
+            if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, projectedRange, mappingBehavior, out var hostDocumentRange))
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            _logger.LogTrace($"Breakpoint span request for ({request.Position.Line}, {request.Position.Character}) = ({hostDocumentRange.Start.Line}, {hostDocumentRange.Start.Character}");
+
+            return new RazorBreakpointSpanResponse()
+            {
+                Kind = RazorLanguageKind.CSharp,
+                Range = hostDocumentRange,
+                HostDocumentVersion = documentVersion
+            };
+        }
+
+        // Internal for testing
+        internal static MappingBehavior GetMappingBehavior(string hostDocumentPath)
+        {
+            if (hostDocumentPath.EndsWith(".cshtml", FilePathComparison.Instance))
```

Comment:
Ah since we're on the server now we can actually check the `DocumentSnapshot.Kind` explicitly. Pretty exciting actually because it also means we can fully disable breakpoints in import files without any extra work + respect any madlads who remapped their cshtml files into Blazor files😄